### DISCUSSION TITLE
How to Run Cron on Laragon ?

### DISCUSSION THREAD

**Discussion by reactmore** [leokhoa/laragon#356] [Q&A] [answered]
hello guys, how to run cron feature on laragon ?

**Comment by Pen-y-Fan** [comment]
[Laragon 6](https://github.com/leokhoa/laragon/releases) has added Cron = cronical - `Menu > Tools > cron > cronical.dat`

**Reply by jennydupuy** [reply]
Hi,

No documentation about this tool. Some examples would be welcome.

**Reply by Pen-y-Fan** [reply, marked as answer]
The documentation is in the README on GitHub: [mgefvert/Cronical](https://github.com/mgefvert/Cronical).

There is also some of documentation and examples in the **cronical.dat** file.

## Example

Open **conical.dat** (`Menu > Tools > cron > cronical.dat`)

Copy the example for every minute and every 5 minutes to the end of the file and remove the comment #

add `>> my_log.txt` to the end of each so the echo is piped to a log file **my_log.txt**

```text
# --- Jobs --------------------------------------------------------------------------

*    *    * * *        cmd /c echo Every minute >> my_log.txt

*/5  *    * * *        cmd /c echo Every five minutes >> my_log.txt
```

Start **terminal** (or cmd or power shell) in admin mode hold the windows key <kbd>Win</kbd> press <kbd>x</kbd> and select  **Terminal (Admin)**

Allow elevated privileges.

Navigate to **cronical** in laragon

E.g. by default:

```shell
cd c:\laragon\bin\cronical
```

To see cronical working, follow the documentation and run in debug mode:

```shell
 .\Cronical.exe --console --debug
```

It will start and the output will be displayed in the terminal. Leave it running for five minutes and both jobs will have run:

```text
20:02:31.213 ***   Cronical booting up
20:02:31.223       Starting Cronical as a console program...
20:02:31.224 ***   Process startup
20:02:31.224       Using definition file cronical.dat
20:02:31.235       - Next job start 12/01/2023 20:03:00 for cmd /c echo Every minute >> my_log.txt
20:02:31.237       - Next job start 12/01/2023 20:05:00 for cmd /c echo Every five minutes >> my_log.txt
20:02:31.238       Config: Run missed jobs on startup = True
20:02:31.238       Config: Check services every       = 15 seconds
20:02:31.239       Config: Terminate cron jobs after  = 3600 seconds
20:02:31.239       2 jobs in job list
20:02:31.240       - Starting boot jobs
20:02:31.241       - Run missed jobs mode - recalculating jobs execution time from last activity...
20:02:31.241       - Next job start 12/01/2023 19:56:00 for cmd /c echo Every minute >> my_log.txt
20:02:31.241       - Next job start 12/01/2023 20:00:00 for cmd /c echo Every five minutes >> my_log.txt
20:02:32.247       - [4] Next job start 12/01/2023 20:03:00 for cmd /c echo Every minute >> my_log.txt
20:02:32.248       - [4] Next job start 12/01/2023 20:05:00 for cmd /c echo Every five minutes >> my_log.txt
20:02:32.248       - [4] Checking services
20:02:32.249       [5] Starting job: cmd /c echo Every minute >> my_log.txt
20:02:32.249       [6] Starting job: cmd /c echo Every five minutes >> my_log.txt
20:02:32.249       [4] Hello! I have 2 upcoming jobs today and 0 services running.
20:02:32.250       - [5] Executing: C:\WINDOWS\SYSTEM32\cmd.exe
20:02:32.250       - [6] Executing: C:\WINDOWS\SYSTEM32\cmd.exe
20:02:32.251       - [5] ...parameters: /c echo Every minute >> my_log.txt
20:02:32.251       - [6] ...parameters: /c echo Every five minutes >> my_log.txt
20:02:32.251       - [6] ...in directory: E:\laragon\bin\cronical
20:02:32.251       - [5] ...in directory: E:\laragon\bin\cronical
20:02:32.259       - [6] Process started, waiting at most 3600 seconds
20:02:32.259       - [5] Process started, waiting at most 3600 seconds
20:02:33.282       - [6] Job finished
20:02:33.282       - [5] Job finished
20:02:47.382       - [7] Checking services
20:03:00.495       - [8] Next job start 12/01/2023 20:04:00 for cmd /c echo Every minute >> my_log.txt
20:03:00.495       [9] Starting job: cmd /c echo Every minute >> my_log.txt
20:03:00.496       - [9] Executing: C:\WINDOWS\SYSTEM32\cmd.exe
20:03:00.496       - [9] ...parameters: /c echo Every minute >> my_log.txt
20:03:00.497       - [9] ...in directory: E:\laragon\bin\cronical
20:03:00.500       - [9] Process started, waiting at most 3600 seconds
20:03:01.525       - [9] Job finished
20:03:02.520       - [7] Checking services
20:03:17.672       - [7] Checking services
20:03:32.821       - [7] Checking services
20:03:47.949       - [4] Checking services
20:04:01.069       - [8] Next job start 12/01/2023 20:05:00 for cmd /c echo Every minute >> my_log.txt
20:04:01.069       [10] Starting job: cmd /c echo Every minute >> my_log.txt
20:04:01.069       - [10] Executing: C:\WINDOWS\SYSTEM32\cmd.exe
20:04:01.069       - [10] ...parameters: /c echo Every minute >> my_log.txt
20:04:01.069       - [10] ...in directory: E:\laragon\bin\cronical
20:04:01.074       - [10] Process started, waiting at most 3600 seconds
20:04:02.092       - [10] Job finished
20:04:03.091       - [4] Checking services
20:04:18.235       - [4] Checking services
20:04:33.380       - [4] Checking services
20:04:48.540       - [4] Checking services
20:05:00.639       - [4] Next job start 12/01/2023 20:06:00 for cmd /c echo Every minute >> my_log.txt
20:05:00.639       - [4] Next job start 12/01/2023 20:10:00 for cmd /c echo Every five minutes >> my_log.txt
20:05:00.639       [11] Starting job: cmd /c echo Every minute >> my_log.txt
20:05:00.640       - [11] Executing: C:\WINDOWS\SYSTEM32\cmd.exe
20:05:00.640       [12] Starting job: cmd /c echo Every five minutes >> my_log.txt
20:05:00.641       - [12] Executing: C:\WINDOWS\SYSTEM32\cmd.exe
20:05:00.641       - [12] ...parameters: /c echo Every five minutes >> my_log.txt
20:05:00.642       - [12] ...in directory: E:\laragon\bin\cronical
20:05:00.640       - [11] ...parameters: /c echo Every minute >> my_log.txt
20:05:00.642       - [11] ...in directory: E:\laragon\bin\cronical
20:05:00.645       - [12] Process started, waiting at most 3600 seconds
20:05:00.648       - [11] Process started, waiting at most 3600 seconds
20:05:01.678       - [11] Job finished
20:05:01.678       - [12] Job finished
20:05:03.670       - [4] Checking services
20:05:18.818       - [4] Checking services
20:05:33.964       - [4] Checking services
20:05:49.114       - [4] Checking services
20:06:01.242       - [4] Next job start 12/01/2023 20:07:00 for cmd /c echo Every minute >> my_log.txt
20:06:01.243       [10] Starting job: cmd /c echo Every minute >> my_log.txt
20:06:01.243       - [10] Executing: C:\WINDOWS\SYSTEM32\cmd.exe
20:06:01.243       - [10] ...parameters: /c echo Every minute >> my_log.txt
20:06:01.244       - [10] ...in directory: E:\laragon\bin\cronical
20:06:01.247       - [10] Process started, waiting at most 3600 seconds
20:06:02.268       - [10] Job finished
20:06:04.272       - [8] Checking services
20:06:19.413       - [8] Checking services
20:06:34.534       - [8] Checking services
20:06:49.674       - [8] Checking services
20:07:00.774       - [7] Next job start 12/01/2023 20:08:00 for cmd /c echo Every minute >> my_log.txt
20:07:00.775       [9] Starting job: cmd /c echo Every minute >> my_log.txt
20:07:00.775       - [9] Executing: C:\WINDOWS\SYSTEM32\cmd.exe
20:07:00.775       - [9] ...parameters: /c echo Every minute >> my_log.txt
20:07:00.775       - [9] ...in directory: E:\laragon\bin\cronical
20:07:00.779       - [9] Process started, waiting at most 3600 seconds
20:07:01.802       - [9] Job finished
20:07:04.818       - [7] Checking services
20:07:19.944       - [7] Checking services
20:07:35.084       - [7] Checking services
20:07:50.238       - [7] Checking services
20:08:01.369       - [8] Next job start 12/01/2023 20:09:00 for cmd /c echo Every minute >> my_log.txt
20:08:01.370       [6] Starting job: cmd /c echo Every minute >> my_log.txt
20:08:01.370       - [6] Executing: C:\WINDOWS\SYSTEM32\cmd.exe
20:08:01.371       - [6] ...parameters: /c echo Every minute >> my_log.txt
20:08:01.371       - [6] ...in directory: E:\laragon\bin\cronical
20:08:01.375       - [6] Process started, waiting at most 3600 seconds
20:08:02.408       - [6] Job finished
20:08:05.411       - [4] Checking services
20:08:20.545       - [4] Checking services
20:08:35.686       - [4] Checking services
20:08:50.819       - [4] Checking services
20:09:00.913       - [7] Next job start 12/01/2023 20:10:00 for cmd /c echo Every minute >> my_log.txt
20:09:00.914       [5] Starting job: cmd /c echo Every minute >> my_log.txt
20:09:00.914       - [5] Executing: C:\WINDOWS\SYSTEM32\cmd.exe
20:09:00.915       - [5] ...parameters: /c echo Every minute >> my_log.txt
20:09:00.915       - [5] ...in directory: E:\laragon\bin\cronical
20:09:00.919       - [5] Process started, waiting at most 3600 seconds
20:09:01.936       - [5] Job finished
20:09:05.956       - [8] Checking services
20:09:21.098       - [8] Checking services
....
```

To exit hold <kbd>ctrl</kbd> and press <kbd>c</kbd>

```text
20:09:24.316 ***   [4] Break signaled, exiting
20:09:24.316 ***   Shutting down
```

- The **Every minute** job has run at the beginning of every minute. 
- The **Every five minutes** job ran on startup and on the 5-minute mark. It was scheduled to run 10 minutes past the hour.

The log file can be opened using notepad:

```shell
notepad my_log.txt
```

Notepad will launch and open **my_log.txt**. The output from the jobs can be seen:

```text
Every five minutes 
Every minute 
Every minute 
Every minute 
Every five minutes 
Every minute 
Every minute 
Every minute 
Every minute 
Every minute 
```

Modify the **cronical.dat** to run the required jobs at the required frequency.

The **README** has more information on how to run jobs as a service. Note the documentation states:

> Start Cronical from an **elevated** console:

For more examples on scheduling cron jobs see the Linux man for [crontab](https://linux.die.net/man/5/crontab)

I hope this helps.

**Comment by HenkBarreveld** [comment]
I wanted to use Cronical the "Laragon-way" (_Laragon has an isolated environment and uses built-in service orchestration, so it doesn’t affect your OS_) and that turned out be surprisingly simple. This what I have now:

Step 1: Open the Laragon menu > Laragon > Procfile > Procfile and add a command line 
`Cron: autorun "Cronical --console >NUL 2>&1" PWD=bin\cronical`

Step 2: restart Laragon (or restart your computer).
Cronical now starts automatically as a background process when Laragon starts. You will not see it, unless you view the Cronical logfile (in `laragon\bin\cronical`) or the Windows Task Manager.

Step 3: Open the Laragon menu > Tools > Cron > cronical.dat and add your cronjobs. For example, for a WordPress website `site.test` you can activate WP_Cron every minute with  a cronjob `* * * * * php C:\laragon\www\site\wp-cron.php`

This works fine, but has two disadvantages:

1. The Laragon menu > Laragon now has at the top an item `Cron* - Autorun`, which you are tempted to click, but shouldn't :-)
2. When you stop Laragon, Cronical keeps running; you have to stop it separately with the Windows Task Manager, or so.

Does anybody have a suggestion how to get around these?

**Reply by Pen-y-Fan** [reply]
For problem 2 you can use the command line to stop Cronical, which can be a batch file on the desktop, see my blog about [https://pen-y-fan.github.io/2023/02/23/how-to-install-mailpit-in-laragon/#how-to-stop-mailpit](How to stop Mailpit)

I wouldn't recommend using Laragon to autostart Cronical, a shortcut on the desktop may be better, or Windows Task Scheduler starts Conical at logon.

**Reply by HenkBarreveld** [reply]
Hello Michael,

Thank you and, indeed, the solution described in _How to stop Mailpit_ is much nicer than using the Windows Task Manager.

In the meantime I found a, for me, even simpler way to handle this. The Procfile command I use, causes Cronical to be run from a Terminal. That terminal is hidden because the command is put between guotes ("); see https://laragon.org/docs/easy-to-extend.htm. Clicking the item `Cron* - Autorun` at the top of the Laragon menu > Laragon will bring the Terminal running Cronical into view. Its window is empty, because the Cronical output is redirected to NUL in the Procfile command. 

Clicking the Close-button of that terminal stops Cronical.

When you have stopped Cronical this way, you can start it again by clicking the same item `Cron* - Autorun` at the top of the Laragon menu > Laragon.

This is fine for me.

By the way, I am curious... for what reason do you think it is better not to have Laragon start Cronical, but to start it manually or with the Task Scheduler?

**Reply by Pen-y-Fan** [reply]
> By the way, I am curious... for what reason do you think it is better not to have Laragon start Cronical, but to start it manually or with the Task Scheduler?

With Mailpit / Mailhog only one instance will run, I'm not sure with Cronical. It may cause problems if multiple instances start running.

**Reply by MotionPhix** [reply]
With that being said, how can I successfully run Laravel schedule commands? I tried

`* * * * * cmd cd "C:\laragon\www\nerd-reports" && cmd "php artisan schedule:run" >> /dev/null 2>&1`

However, that isn't doing anything. It runs every minute, but nothing is happening. I had to add those quotes to probably make it work but to no avail. The actual command is 
`* * * * * cd C:\laragon\www\nerd-reports && php artisan schedule:run >> /dev/null 2>&1`

**Reply by Pen-y-Fan** [reply]
I'm not sure multi-line command (with &&) will work. Try a single line:

```text
* * * * * php C:\laragon\www\nerd-reports\artisan schedule:run
```

Or create a multi-line batch file in C:\laragon\www\nerd-reports\ and call it, say **run-schedule-run.bat**, which contains your `php artisan schedule:run` command:

```cmd
* * * * *        cmd /c C:\laragon\www\nerd-reports\run-schedule-run.bat >> my_log.txt
```

It may need a bit of trial and error to get working. Please let us know how you get on.

**Reply by AWqxKAWERbXo** [reply]
Hey guys,

I wonder how I even start the service. Trying to execute Cronical.exe results in this error message:

![image](https://github.com/user-attachments/assets/6e6a4f06-c1f1-47c8-ab66-9cf15d6fd9d5)

It took me a little bit to find out that I can't just run `cronical.exe --install` from PowerShell or as a regular user. I needed to start it as an admin in the classic CMD.

![image](https://github.com/user-attachments/assets/aabda7ff-d868-4ee7-83c1-d3687755505c)

Afterward, I ran `net start cronical`. I didn't fully understand how can I see if cronical is running as a service or not and how to make it start automatically.

But I also get this error:
```
2025-01-06 22:19:01.609 WARN  [4] File C:\laragon\bin\cronical\php does not seem to exist
```

Why doesn't it recognize php as it should (Environment Variables?)

```log
2025-01-06 22:19:01.584 ***   [4] Process startup
2025-01-06 22:19:01.584       [4] Using definition file cronical.dat
2025-01-06 22:19:01.609 WARN  [4] File C:\laragon\bin\cronical\php does not seem to exist
2025-01-06 22:19:01.609       [4] Config: Run missed jobs on startup = True
2025-01-06 22:19:01.609       [4] Config: Check services every       = 15 seconds
2025-01-06 22:19:01.609       [4] Config: Terminate cron jobs after  = 3600 seconds
2025-01-06 22:19:01.609       [4] 1 jobs in job list
2025-01-06 22:19:02.628       [4] Hello! I have 1 upcoming jobs today and 0 services running.
2025-01-06 22:20:01.038       [7] Starting job: php C:\laragon\www\my-project.de\bin\console cronjob:run
2025-01-06 22:20:01.050 ERR   [7] Failed to start process 'php C:\laragon\www\my-project.de\bin\console cronjob:run': Das System kann die angegebene Datei nicht finden
2025-01-06 22:21:00.517       [8] Starting job: php C:\laragon\www\my-project.de\bin\console cronjob:run
2025-01-06 22:21:00.517 ERR   [8] Failed to start process 'php C:\laragon\www\my-project.de\bin\console cronjob:run': Das System kann die angegebene Datei nicht finden
2025-01-06 22:22:01.073       [9] Starting job: php C:\laragon\www\my-project.de\bin\console cronjob:run
2025-01-06 22:22:01.073 ERR   [9] Failed to start process 'php C:\laragon\www\my-project.de\bin\console cronjob:run': Das System kann die angegebene Datei nicht finden
2025-01-06 22:23:00.545       [10] Starting job: php C:\laragon\www\my-project.de\bin\console cronjob:run
2025-01-06 22:23:00.547 ERR   [10] Failed to start process 'php C:\laragon\www\my-project.de\bin\console cronjob:run': Das System kann die angegebene Datei nicht finden
```
Can someone assist me jump starting the cron?

**Comment by HenkBarreveld** [comment]
You are right about this: with autorun enabled for Cronical, restarting Laragon a few times causes as many copies of Cronical to be started as well. I tested this with `--debug` on: all copies are running nicely together; so to see independent of each other - except that they share the same log file.

For me this is not a problem, because I use Cronical only on WordPress websites, to run wp-cron.php periodically. That script protects itself against re-entry wile it is still running and the only requirement is that it runs often enough. Once per minute is perfect; more frequently is unnecessary but does not harm.
